### DISCUSSION TITLE
How do self-ordering Kiosks softwares work actually? are they Web or program? What programming languages?

### DISCUSSION THREAD

**Discussion by kedybogan** [community/community#57922] [Other Feature Feedback, Questions, & Ideas]
### Select Topic Area

Question

### Body

Hi, I am currently exploring the development of kiosk software and need some guidance regarding the appropriate programming languages and frameworks to use. Specifically, I am interested in understanding the technologies behind popular kiosk software systems like McDonald's Kiosks or GRUBBRR.
I need guidance, help please feel free to share any information with me.

Initially, I thought that developing a Linux Ubuntu Kiosk and implementing a Java program would suffice. However, I am now feeling quite confused, particularly when it comes to the user interface (UI) aspect. I have looked into Java Swing, but it doesn't seem to align with what I see in these kiosk systems.

My main questions are:

What programming languages and frameworks are commonly used in the development of kiosk software?

How do kiosk software systems like McDonald's Kiosks or GRUBBRR handle their user interfaces? Is it through Ruby or other web-based technologies?

Does Ruby typically imply web development? In other words, are these kiosk machines leveraging web technologies instead of traditional PC programs?

I would greatly appreciate any insights, advice, or recommended resources to help me better understand the technology stack used in kiosk software development.

So what they exactly do, and how does Kiosks are software work? Can I make great UI software in Java or is it possible and more effective if I also learn to combine Ruby etc? Are their software work via the web instead of a .exe program? It is so confusing.

Thank you in advance for your help!

**Comment by perk-bevis** [comment]
Kiosk software development can involve a variety of programming languages and frameworks depending on the specific requirements and platforms used. While there is no one-size-fits-all answer, I can provide you with some common technologies used in kiosk software development:

Web technologies: Many modern kiosk systems utilize web-based technologies to develop their user interfaces. This typically involves a combination of HTML, CSS, and JavaScript to create interactive web-based UIs. Frameworks such as React, Angular, or Vue.js may be used to enhance the development process and create more sophisticated interfaces.

Backend programming languages: The backend of a kiosk software system handles the processing of data and communication with external systems. Common programming languages for backend development include Python, Java, C#, and Ruby. These languages can be used to handle business logic, database interactions, and API integrations.

Embedded systems: Some kiosk software is developed specifically for embedded systems, such as kiosks running on customized hardware or operating systems. In such cases, programming languages like C, C++, or Rust may be used to interact directly with the hardware and optimize performance.

Now, let's focus on the examples you mentioned. While I don't have specific information about the technologies used in McDonald's Kiosks or GRUBBRR, it's common for modern kiosk systems to leverage web technologies. They may use a combination of web-based UIs and backend services to provide functionality and communicate with other systems.

Regarding Ruby, it is a programming language that is often associated with web development due to its popular web framework called Ruby on Rails. However, Ruby can be used for various purposes beyond web development, including scripting, automation, and backend services. While it's possible that Ruby could be used in some kiosk software systems, it's not the only option, and the choice of programming language depends on the specific requirements of the project.

In terms of creating great UI software, Java is a popular language with a wide range of UI frameworks and libraries available. JavaFX, for example, is a powerful UI framework that can be used to develop rich and interactive user interfaces. Combining Java with other technologies like web-based frameworks can provide additional flexibility and capabilities, depending on your project's requirements.

Ultimately, the choice of programming languages and technologies depends on factors such as the specific requirements, the target platform, hardware limitations, development resources, and the desired user experience. To gain a deeper understanding of kiosk software development, I recommend exploring relevant online resources, tutorials, and documentation for specific programming languages and frameworks that interest you. This will help you grasp the concepts and best practices involved in building kiosk software systems.

**Reply by kedybogan** [reply]
Thank you for your reply. I am still worried about if Java can handle the UI. I have never seen a real program or course that made on Java Swing as I researched. All of them like just beginning / learning projects so that made me think it is not really suitable for that professional purpose software.
I think I will learn web stuff, front-end languages HTML, CSS, Javascript, Vue and what it could help anyways and then I'll work on combining Java back-end with the web. I am trying to research a lot but there aren't really a lot of sources for it.

So, by the end of the day, I think the software will be web-based and I think it will not result me any problem and work seamlessly in every part in the future.
